### PR TITLE
Upgrade to Docsy 0.5.1 via NPM module

### LINKED ISSUE #32905: Bootstrap and Font Awesome (as part of Docsy) are deprecated
**This is a Bug Report**

**Problem:**
https://github.com/google/docsy/discussions/950 marks that **within the context of the Docsy** theme:
- Bootstrap as a Git submodule is deprecated
- Font Awesome as a Git submodule is deprecated

Docsy still use these things, but the recommended. mechanism for that is no longer Git submodules.

**Proposed Solution:**
come up with a new way to build the site via Netlify and 
- this might involve adding `npm` to the story, or it might not
- this might involve using Hugo modules

### LINKED ISSUE #48806: There are two <script> elements referencing jQuery
There should be only one `<script>` element for jQuery, but at the moment jQuery is referenced twice:

```console
$ find layouts -type f -exec grep -nH -ie jquery  {} +
layouts/partials/scripts.html:1:<script src="/js/jquery-3.6.0.min.js" integrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" crossorigin="anonymous"></script>
layouts/partials/head.html:76:<script src="/js/jquery-3.6.0.min.js" intregrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" crossorigin="anonymous"></script>
```

```console
$ curl -s https://kubernetes.io/docs/home/ | tr '<' '\n' | grep -i '^script' | grep -i 'jquery' 
script src=/js/jquery-3.6.0.min.js intregrity=sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK crossorigin=anonymous>
script src=/js/jquery-3.6.0.min.js integrity=sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK crossorigin=anonymous>
```

### LINKED ISSUE #48807: Upgrade to Docsy 0.5.1
- We'll be building upon the successful integration of: #48721
- This is a step of #44002
- Partly contributes to #41171
- Upgrade to 0.5.1 from 0.3.0, because 0.5.1 is the first official version that supports use of Docsy via an NPM module.

Issues found and to be resolved in scope of this work:

- [ ] #32905
- [ ] #48806
- [ ] #48808
- [ ] #48811. Actually, all v3-v5 IDs should be updated to use the [v6 icon](https://fontawesome.com/icons)'s canonical name
- [ ] #48813

Note that @sftim did some preliminary exploration:

- #48722
- #48723
- Etc

Resources:

- [Changelog#v0.5.1](https://www.docsy.dev/project/changelog/#051)
- https://github.com/google/docsy/compare/v0.3.0...v0.5.1

/assign
/cc @nate-double-u

### LINKED ISSUE #48808: Named code-block elements show unnecessary block padding
For example, visit https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#topology-spread-constraint-examples and you'll currently see:

> image

There shouldn't be whitespace around the grey code block -- i.e., the code block should extend to the border of it's parent container.

### LINKED ISSUE #48811: Use canonical FontAwesome v6 icon names, upgrading from v3-v5
**Edit**: Actually, all v3-v5 IDs should be updated to use the [v6 icon](https://fontawesome.com/icons)'s canonical name, such as the following for GitHub once Docsy is upgraded to 0.5.1+ (#48807):

```html
<i class="fa-brands fa-github"></i>
```

---

Visit https://kubernetes.io/docs/concepts/windows/intro/ and you'll currently see the following page-meta links:

> Screen Shot 2024-11-22 at 17 21 42

Notice that the `<i class="fa fa-icon-circle fa-fw"></i>` used to generate this icon doesn't render. The use of this icon was introduced via #29877, but even in that PR's preview the icon doesn't render, see: https://deploy-preview-29877--kubernetes-io-main-staging.netlify.app/docs/setup/production-environment/container-runtimes/#container-runtimes

> image

### FA v3 icon!

It took some digging, but I finally found it! That is 3 major versions back 😅. It was a FA v3 icon: https://fontawesome.com/v3/icon/circle

> image

### Replacement

The replacement seems to be https://fontawesome.com/icons/circle?f=classic&s=solid:

```html
<i class="fa-solid fa-circle"></i>
```

### LINKED ISSUE #48813: Tab panes have formatting issues
Visit https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#install-kubectl-binary-with-curl-on-linux and you'll see:

> image

Notice that there is no full bounding box.

### PR DESCRIPTION
**Pull request by chalin**
- Fixes #48807
- Contributes to:
  - #41171
  - #44002
- Upgrades to Docsy 0.5.1 as an NPM module.
  - Fetches FontAwesome and Bootstrap via NPM, so I believe (if I understand the intent) that this 
    fixes #32905
  - Upgrades FontAwesome to v6, and so upgrades all icon names, and so
    fixes #48811
- Drops the git submodule reference to Docsy 0.3.0
- Ensures that container commands continue to work
- Updates the makefile
- Switches to using `template "partials/page-description.html"` to generate HTML header page meta descriptions. (This template has continued to be improved, with the most recent update being made in Docsy 0.11.0.)
- Drops the `static` Bootstrap resources, keeping the one needed by the API pages.

Also addresses and closes the following:

- Fixes #48806
  ```console
  $ curl -s https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/ | tr '<' '\n' | grep -i '^script' | grep -i 'jquery' 
  script src="/js/jquery-3.6.0.min.js" intregrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" crossorigin="anonymous">
  ```
- Fixes #48808
- Fixes #48813

## Previews

- https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/ - e.g., check the footer
- https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/docs/concepts/windows/ for the page-meta icons, including the 3rd-party one
- https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/configmap/#using-configmaps-as-environment-variables - for code-blocks, including named ones with a copy-code button, as well as tooltips
- https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/docs/tasks/tools/install-kubectl-linux/#install-kubectl-binary-with-curl-on-linux for tabbed panes, e.g.:
  > image
- https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/docs/contribute/docs/#getting-started for Mermaid diagrams

## Checks I've done

I've checked that the following render correctly, feel free to double check:

- **Icons** (now that we're using FA v6) render properly in the
  - Footer
  - Doc pages, page-meta-links section
  - Blog pages, page-meta and other links
  - More rarely used icons like the third-party icon (in page-meta), for example see Concepts > Windows
- **Mermaid** diagrams continue to render correctly
- **Tooltips** are still working, demonstrating that the new way to fetch and use Bootstrap CSS and JS is working
- Code-blocks and figures render correctly

## Followup changes

This PR makes minimal changes. Consider following up with updates to:

- API reference generation code to catch up to the FontAwesome & Bootstrap versions now used by the rest of the website (via Docsy).
- Dockerfile, to switch to `docker.io/library/debian:bookworm` as @sftim proposed in #48722
- #48814
- Btw, stylistically, I think that we should drop the `border` on the `.mermaid` class

Related:

- #48723

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -56,7 +56,7 @@ non-production-build: module-check ## Build the non-production site, which adds
 	GOMAXPROCS=1 hugo --cleanDestinationDir --enableGitInfo --environment nonprod
 
 serve: module-check ## Boot the development server.
-	hugo server --buildFuture --environment development
+	hugo serve --minify --buildFuture --environment development --disableFastRender --renderToMemory
```
Comment (sftim) [inline Makefile:59]: The change to memory rendering is probably OK, but we should check it works on svelte PCs, and I think it belongs in a separate PR.

File: Makefile [hunk 2]
```diff
@@ -107,7 +107,7 @@ container-build: module-check
 container-serve: module-check ## Boot the development server using container.
 	$(CONTAINER_RUN) --cap-drop=ALL --cap-add=AUDIT_WRITE --read-only \
 		--mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 -p 1313:1313 $(CONTAINER_IMAGE) \
-		hugo server --buildFuture --environment development --bind 0.0.0.0 --destination /tmp/public --cleanDestinationDir --noBuildLock
+		hugo serve --minify --renderToMemory --buildFuture --environment development --bind 0.0.0.0 --destination /tmp/public --cleanDestinationDir --noBuildLock
```
Comment (sftim) [inline Makefile:110]: Does this mean that you needed to run `npm ci` outside the container first? If so:
- let's document that
- let's get a mini security review

File: Makefile
Comment (sftim) [inline Makefile:1]:
```console
$ CONTAINER_ENGINE=podman nice -n 5 make container-serve
"podman" run --rm --interactive --tty --volume "/home/tim/Documents/Repositories/kubernetes/website:/src:ro,Z" --cap-drop=ALL --cap-add=AUDIT_WRITE --read-only \
	--mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 -p 1313:1313 gcr.io/k8s-staging-sig-docs/k8s-website-hugo:v0.133.0-9a7409a1d2aa \
	hugo serve --minify --renderToMemory --buildFuture --environment development --bind 0.0.0.0 --destination /tmp/public --cleanDestinationDir --noBuildLock
WARN[0000] Error validating CNI config file /home/tim/.config/cni/net.d/hands-on_default.conflist: [failed to find plugin "dnsname" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin]]
Hugo provides its own webserver which builds and serves the site.
While hugo server is high performance, it is a webserver with limited options.

The `hugo server` command will by default write and serve files from disk, but
you can render to memory by using the `--renderToMemory` flag. This can be
faster in some cases, but it will consume more memory.

By default hugo will also watch your files for any changes you make and
automatically rebuild the site. It will then live reload any open browser pages
and push the latest content to them. As most Hugo sites are built in a fraction
of a second, you will be able to save and see your changes nearly instantly.

Usage:
  hugo server [command] [flags]
  hugo server [command]

Aliases:
  server, serve

Available Commands:
  trust       Install the local CA in the system trust store.

Flags:
      --appendPort               append port to baseURL (default true)
  -b, --baseURL string           hostname (and path) to the root, e.g. https://spf13.com/
      --bind string              interface to which the server will bind (default "127.0.0.1")
  -D, --buildDrafts              include content marked as draft
  -E, --buildExpired             include expired content
  -F, --buildFuture              include content with publishdate in the future
      --cacheDir string          filesystem path to cache directory
      --cleanDestinationDir      remove files from destination not found in static directories
  -c, --contentDir string        filesystem path to content directory
      --disableBrowserError      do not show build errors in the browser
      --disableFastRender        enables full re-renders on changes
      --disableKinds strings     disable different kind of pages (home, RSS etc.)
      --disableLiveReload        watch without enabling live browser reload on rebuild
      --enableGitInfo            add Git revision, date, author, and CODEOWNERS info to the pages
      --forceSyncStatic          copy all files when static is changed.
      --gc                       enable to run some cleanup tasks (remove unused cache files) after the build
  -h, --help                     help for server
      --ignoreCache              ignores the cache directory
  -l, --layoutDir string         filesystem path to layout directory
      --liveReloadPort int       port for live reloading (i.e. 443 in HTTPS proxy situations) (default -1)
      --minify                   minify any supported output format (HTML, XML etc.)
  -N, --navigateToChanged        navigate to changed content file on live browser reload
      --noBuildLock              don't create .hugo_build.lock file
      --noChmod                  don't sync permission mode of files
      --noHTTPCache              prevent HTTP caching
      --noTimes                  don't sync modification time of files
      --panicOnWarning           panic on first WARNING log
      --poll string              set this to a poll interval, e.g --poll 700ms, to use a poll based approach to watch for file system changes
  -p, --port int                 port on which the server will listen (default 1313)
      --pprof                    enable the pprof server (port 8080)
      --printI18nWarnings        print missing translations
      --printMemoryUsage         print memory usage to screen at intervals
      --printPathWarnings        print warnings on duplicate target paths etc.
      --printUnusedTemplates     print warnings on unused templates.
      --renderSegments strings   named segments to render (configured in the segments config)
      --renderStaticToDisk       serve static files from disk and dynamic files from memory
      --templateMetrics          display metrics about template executions
      --templateMetricsHints     calculate some improvement hints when combined with --templateMetrics
  -t, --theme strings            themes to use (located in /themes/THEMENAME/)
      --tlsAuto                  generate and use locally-trusted certificates.
      --tlsCertFile string       path to TLS certificate file
      --tlsKeyFile string        path to TLS key file
      --trace file               write trace to file (not useful in general)
  -w, --watch                    watch filesystem for changes and recreate as needed (default true)

Global Flags:
      --clock string               set the clock used by Hugo, e.g. --clock 2021-11-06T22:30:00.00+09:00
      --config string              config file (default is hugo.yaml|json|toml)
      --configDir string           config dir (default "config")
      --debug                      debug output
  -d, --destination string         filesystem path to write files to
  -e, --environment string         build environment
      --ignoreVendorPaths string   ignores any _vendor for module paths matching the given Glob pattern
      --logLevel string            log level (debug|info|warn|error)
      --quiet                      build in quiet mode
  -M, --renderToMemory             render to memory (mostly useful when running the server)
  -s, --source string              filesystem path to read files relative from
      --themesDir string           filesystem path to themes directory
  -v, --verbose                    verbose output

Use "hugo server [command] --help" for more information about a command.

Error: command error: cannot use --renderToMemory with --destination
make: *** [Makefile:108: container-serve] Error 1
```

:warning: I get an error trying to run a local container preview.

Comment (sftim) [reply Makefile:1]: If I drop `--renderToMemory`, the error changes to:
```console
Error: command error: failed to load modules: module "docsy" not found in "/src/node_modules/docsy"; either add it as a Hugo Module or store it in "/src/node_modules".: module does not exist
make: *** [Makefile:108: container-serve] Error 1
```

File: Makefile (hunk 2 above)
Comment (sftim) [reply Makefile:110]: I'd really prefer to keep the setup where `npm ci` runs within the container image build step. I hope we can.

:thought_balloon: to make that work right, we should really change `scripts/hash-files.sh` to include `package-lock.json`

File: Makefile (hunk 1 above)
Comment (chalin) [reply Makefile]: Agreed. Change to `serve` target reverted.

File: layouts/partials/head.html [hunk 3]
```diff
@@ -55,9 +55,9 @@
 {{ partial "css.html" . }}
 <!-- Content for social media sharing previews -->
 <!-- Facebook uses the og: stuff, while Twitter and others use twitter: -->
-<meta name="description" content="{{ if .Params.description }}{{ .Params.description }}{{ else }}{{ .Summary }}{{ end }}">
-<meta property="og:description" content="{{ if .Params.description }}{{ .Params.description }}{{ else }}{{ .Summary }}{{ end }}">
-<meta name="twitter:description" content="{{ if .Params.description }}{{ .Params.description }}{{ else }}{{ .Summary }}{{ end }}">
+<meta name="description" content="{{ template "partials/page-description.html" . }}">
+<meta property="og:description" content="{{ template "partials/page-description.html" . }}">
+<meta name="twitter:description" content="{{ template "partials/page-description.html" . }}">
```
Comment (sftim) [inline layouts/partials/head.html:60]: Nice

File: Makefile (hunk 2 above)
Comment (chalin) [reply Makefile]: I've reverted this change since it isn't essential to the 0.5.1 upgrade.

All good comments. Could you create a separate issue to track those, since they're outside of the scope of this PR. Thanks!

File: netlify.toml [hunk 4]
```diff
@@ -4,10 +4,10 @@
 # DO NOT REMOVE THIS (contact @kubernetes/sig-docs-leads)
 publish = "public"
 functions = "functions"
-command = "git submodule update --init --recursive --depth 1 && make non-production-build && npx -y pagefind --site public"
+command = "make non-production-build && npx -y pagefind --site public"
 
 [build.environment]
-NODE_VERSION = "20.17.0"
+NODE_VERSION = "20.11.0"
```
Comment (sftim) [inline netlify.toml:10]: Calling out a downgrade here

File: Makefile
Comment (chalin) [reply Makefile:1]: Thanks for testing the container commands. Since I've reverted all changes to the `Makefile`, please create a separate issue so that we can track improvements to the Makefile and Dockerfile.

File: netlify.toml (hunk 4 above)
Comment (chalin) [reply netlify.toml]: Oh, that shouldn't be there. Reverting this change.

**Comment by sftim** [comment]
Try running `mv node_modules node_modules_moved_aside` before trying a local preview in a container. That should show up the snags, if there are any.

**Comment by sftim** [comment]
Hope this is OK
/retitle [WIP] Upgrade to Docsy 0.5.1 via NPM module

**Comment by chalin** [comment]
@sftim wrote via https://github.com/kubernetes/website/issues/48811#issuecomment-2495582965:

> You won't like this @chalin but the `<i class="fa fa-icon-circle fa-fw"></i>` was a **bodge** to get a correctly-indented second line with no icon.
> 
> That's the original presentation intent, so bear that in mind when considering what we'd like to preserve…

Yes, here's the fix I propose: https://github.com/kubernetes/website/pull/48812/commits/a65258e1e561d468f33ee1b8580585a7bf6c8756. It looks like this:

> image

which you can see via https://deploy-preview-48812--kubernetes-io-main-staging.netlify.app/docs/concepts/windows/.

**Comment by chalin** [comment]
> Hope this is OK /retitle [WIP] Upgrade to Docsy 0.5.1 via NPM module

@sftim - with the latest changes, this is ready for reivew -- and so not WIP. The Docker & build script issues being orthogonal to the Docsy upgrade.

**Review by Okabe-Junya — COMMENTED** [review]
Great work, thank you chalin!!
LGTM

**Comment by chalin** [comment]
> If this breaks local previewing in a container, we can't merge it.

@sftim - no, the simple change to the Docker file in this PR does not introduce any further breakages to local containerized build that might not have been there before.

Thanks @Okabe-Junya.
All: anything further to be done here?

**Comment by sftim** [comment]
If you don't run `npm` first, the effect of this PR is to break container previewing (you never previously had to run `npm`).

```
Error: command error: failed to load modules: module "docsy" not found in "/src/node_modules/docsy"; either add it as a Hugo Module or store it in "/src/node_modules".: module does not exist
```

File: README.md [hunk 5]
```diff
@@ -30,22 +30,6 @@ git clone https://github.com/kubernetes/website.git
...
-### Linux / other Unix
-
-```bash
-# fetch submodule dependencies
-make module-init
-```
-
 ## Running the website using a container
 
 To build the site in a container, run the following:
```
Comment (nate-double-u) [inline README.md:35]: we may want to update these instructions.

Currently, from a fresh repo, starting with `make module-init` fails with
```
Initializing submodules...
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
fatal: remote error: upload-pack: not our ref 096a94d6e7d0d04e41356f21233498c3c4e31699
fatal: Fetched in submodule path 'api-ref-generator', but it did not contain 096a94d6e7d0d04e41356f21233498c3c4e31699. Direct fetching of that commit failed.
make: *** [module-init] Error 128
```

I can't get past this for running this with a container. I thought maybe I need to do an npm install or npm ci first, but neither of those work from a clean pull either.

Comment (nate-double-u) [reply README.md:35]: I'm looking into my own setup here as I'm not sure this PR should cause this error.

**Comment by chalin** [comment]
> Try running `mv node_modules node_modules_moved_aside` before trying a local preview in a container. That should show up the snags, if there are any.

@sftim - oh, you're right! Sorry for having missed that earlier. I'm looking into a solution.

**Comment by chalin** [comment]
@sftim @nate-double-u et al. - I've had to restore the containerized call to `npm ci` that was removed via #48741. Now both containerized build and serve will run `npm ci` (within the container) before building and serving respectively. PTAL

```console
$ rm -Rf node_modules
$ make container-build
mkdir -p node_modules
"docker" run --rm --interactive --tty ... npm ci
npm warn deprecated popper.js@1.16.1: ...
Run `npm audit` for details.
mkdir -p public
"docker" run --rm --interactive --tty ... \
                hugo --cleanDestinationDir --buildDrafts --buildFuture --environment preview --noBuildLock
Start building sites … 
hugo v0.133.0+extended linux/amd64 BuildDate=unknown


                   |  EN  | BN  | ZH-CN | FR  | DE  | HI  | ID  | IT  | JA  | KO  | PL  | PT-BR | RU  | ES  | UK  | VI   
-------------------+------+-----+-------+-----+-----+-----+-----+-----+-----+-----+-----+-------+-----+-----+-----+------
  Pages            | 2240 | 236 |  1835 | 380 | 210 | 136 | 337 |  92 | 579 | 617 |  71 |   367 | 195 | 319 |  95 |  79  
  Paginator pages  |   61 |   0 |    23 |   0 |   0 |   0 |   0 |   0 |   2 |   0 |   0 |     0 |   0 |   0 |   0 |   0  
  Non-page files   |  737 | 603 |   616 |  84 |  83 |  75 |  80 |  40 | 527 |  86 |  74 |    86 |  80 |  83 |  82 |  75  
  Static files     |  829 | 829 |   829 | 829 | 829 | 829 | 829 | 829 | 829 | 829 | 829 |   829 | 829 | 829 | 829 | 829  
  Processed images |   14 |   0 |     1 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |     0 |   0 |   0 |   0 |   0  
  Aliases          |    8 |   1 |     6 |   2 |   1 |   0 |   1 |   1 |   3 |   3 |   0 |     3 |   4 |   2 |   1 |   0  
  Cleaned          |    0 |   0 |     0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |     0 |   0 |   0 |   0 |   0  

Total in 101228 ms
```

File: Makefile [hunk 6]
```diff
@@ -96,15 +96,23 @@ docker-push: ## Build a multi-architecture image and push that into the registry
...
 	rm Dockerfile.cross
 
-container-build: module-check
+container-npm-ci:
+	mkdir -p node_modules
+	$(CONTAINER_RUN) --read-only \
+		--mount type=bind,source=$(CURDIR)/node_modules,target=/src/node_modules \
+		--mount type=tmpfs,destination=/var/hugo/.npm,tmpfs-mode=01777 \
+		--mount type=tmpfs,destination=/var/hugo/.cache,tmpfs-mode=01777 \
+		$(CONTAINER_IMAGE) npm ci
```
Comment (sftim) [inline Makefile:105]: I'd really prefer to bake the NPM modules into the container image.

(the approach in this PR it also doesn't work for me locally using either Docker or Podman; I'm sure I could get it to work, but baking in a correct `/src/node_modules` would be easier for people)

Eventually we'd like to provide a PowerShell alternative for the `Makefile`. If we can keep the complexity inside the container, that's so much better.

I'll see if I can find a contributor willing to work on reshaping this part for you @chalin.

Comment (chalin) [reply Makefile]: > I'd really prefer to bake the NPM modules into the container image.

Ok. Done, PTAL.

File: Makefile [hunk 7]
```diff
@@ -12,6 +12,22 @@ CONTAINER_IMAGE   = $(IMAGE_REGISTRY)/k8s-website-hugo:v$(HUGO_VERSION)-$(IMAGE_
...
+	--mount type=bind,source=$(CURDIR)/content,target=/src/content,readonly \
+	--mount type=bind,source=$(CURDIR)/data,target=/src/data,readonly \
+	--mount type=bind,source=$(CURDIR)/i18n,target=/src/i18n,readonly \
+	--mount type=bind,source=$(CURDIR)/layouts,target=/src/layouts,readonly \
+	--mount type=bind,source=$(CURDIR)/public,target=/src/public \
+	--mount type=bind,source=$(CURDIR)/resources,target=/src/resources \
+	--mount type=bind,source=$(CURDIR)/static,target=/src/static,readonly \
+	--mount type=bind,source=$(CURDIR)/themes,target=/src/themes,readonly \
+	--mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 \
+	--mount type=bind,source=$(CURDIR)/hugo.toml,target=/src/hugo.toml,readonly
```
Comment (sftim) [inline Makefile:30]: The switch from "the build happens in a container" to "the app can write to the repo" is possibly going to break things; it does for me.

I'm going to see if any of these can be tmpfs (eg `/src/public`) rather than a mount in.

File: Makefile [hunk 8]
```diff
@@ -12,6 +12,22 @@ CONTAINER_IMAGE   = $(IMAGE_REGISTRY)/k8s-website-hugo:v$(HUGO_VERSION)-$(IMAGE_
...
+	--mount type=bind,source=$(CURDIR)/archetypes,target=/src/archetypes,readonly \
+	--mount type=bind,source=$(CURDIR)/assets,target=/src/assets,readonly \
+	--mount type=bind,source=$(CURDIR)/content,target=/src/content,readonly \
+	--mount type=bind,source=$(CURDIR)/data,target=/src/data,readonly \
+	--mount type=bind,source=$(CURDIR)/i18n,target=/src/i18n,readonly \
+	--mount type=bind,source=$(CURDIR)/layouts,target=/src/layouts,readonly \
+	--mount type=bind,source=$(CURDIR)/public,target=/src/public \
+	--mount type=bind,source=$(CURDIR)/resources,target=/src/resources \
+	--mount type=bind,source=$(CURDIR)/static,target=/src/static,readonly \
+	--mount type=bind,source=$(CURDIR)/themes,target=/src/themes,readonly \
```
Comment (sftim) [inline Makefile:28]: Does this directory exist? I'm not sure it does.

File: Makefile (hunk 7 above)
Comment (chalin) [reply Makefile:30]: Are you referringt to writing to `public`? That is how it is already handled in `main`: https://github.com/kubernetes/website/blob/a371603439fde7d34e329e3f88e12e7c5fba5645/Makefile#L99-L104

Can you not build `main`? If you can't run a containerized build from main, then we should create a separate issue to track and fix that. Specifically what is failing?

File: Makefile (hunk 8 above)
Comment (chalin) [reply Makefile:28]: No, that was future-proofing. I'll take it out.

Comment (sftim) [inline Makefile:28]: This works for me, locally:
```suggestion
	--mount type=bind,source=$(CURDIR)/.git,target=/src/.git,readonly \
	--mount type=bind,source=$(CURDIR)/archetypes,target=/src/archetypes,readonly \
	--mount type=bind,source=$(CURDIR)/assets,target=/src/assets,readonly \
	--mount type=bind,source=$(CURDIR)/content,target=/src/content,readonly \
        --mount type=bind,source=$(CURDIR)/data,target=/src/data,readonly \
	--mount type=bind,source=$(CURDIR)/i18n,target=/src/i18n,readonly \
	--mount type=bind,source=$(CURDIR)/layouts,target=/src/layouts,readonly \
	--mount type=tmpfs,destination=/src/public,tmpfs-mode=01777 \
	--mount type=tmpfs,destination=/src/resources,tmpfs-mode=01777 \
	--mount type=bind,source=$(CURDIR)/static,target=/src/static,readonly \
	--mount type=bind,source=$(CURDIR)/hugo.toml,target=/src/hugo.toml,readonly \
	--mount type=tmpfs,destination=/tmp,tmpfs-mode=01777
```

Comment (chalin) [reply Makefile]: I made the switch to declaring tmpfs for public and resources (and pushed the commit so you can test too), but I get an error locally:

```console
$ make container-build
mkdir -p public
"docker" run --rm --interactive --tty --read-only --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/.git,target=/src/.git,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/archetypes,target=/src/archetypes,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/assets,target=/src/assets,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/content,target=/src/content,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/data,target=/src/data,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/i18n,target=/src/i18n,readonly --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/layouts,target=/src/layouts,readonly --mount type=tmpfs,destination=/src/public,tmpfs-mode=01777 --mount type=tmpfs,destination=/src/resources,tmpfs-mode=01777 --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/static,target=/src/static,readonly --mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 --mount type=bind,source=/Users/chalin/git/lf/kubernetes/kubernetes.io/hugo.toml,target=/src/hugo.toml,readonly gcr.io/k8s-staging-sig-docs/k8s-website-hugo:v0.133.0-7886d9227c22 \
                hugo --cleanDestinationDir --buildDrafts --buildFuture --environment preview --noBuildLock
Start building sites … 
hugo v0.133.0+extended linux/amd64 BuildDate=unknown

Total in 1568 ms
Error: error copying static files: chtimes /src/public: operation not permitted
```

I'm investigating further, but wanted you to mention that I'm getting an error.

Comment (chalin) [reply Makefile]: Tthe solution for the containerized build is to use ` --destination /tmp/public`, at least that works for me now. Can you give it a try @sftim, I just pushed the change.

**Review by sftim — COMMENTED** [review]
Informally LGTM

Local container previewing works (tested in Podman only, for now)

File: layouts/partials/scripts.html [hunk 9]
```diff
@@ -1,26 +1,32 @@
...
+  <script src="{{ $js.RelPermalink }}"></script>
+{{ end -}}
+
+{{ if .Site.Params.prism_syntax_highlighting -}}
+  <script src='{{ "js/prism.js" | relURL }}'></script>
+{{ else if false -}}
+  {{ $c2cJS := resources.Get "js/click-to-copy.js" -}}
+  {{ if hugo.IsProduction -}}
+    {{ $c2cJS = $c2cJS | minify | fingerprint -}}
+  {{ end -}}
```
Comment (sftim) [inline layouts/partials/scripts.html:25]: (aside) I wonder if we should use this in place of the click-to-copy code we load from the body-end hook?

**Review by sftim — COMMENTED** [review]
/lgtm

File: Makefile (hunk 7 above)
Comment (chalin) [reply Makefile]: It's too bad that writing to `public` doesn't work for you. It would give folks doing containerized builds the ability to see the generated site files.

**Comment by chalin** [comment]
> See [#48812 (comment)](https://github.com/kubernetes/website/pull/48812#issuecomment-2527893965) for a change I recommend (squashing before we merge it)

Thanks for the LGTM @sftim. Hopefully this will get merged soon. I was hoping to keep the individual commits in this PR because some of the intermediate steps contain solution alternatives that I'd like to reuse elsewhere. The commits will be squashed when the PR gets merged right?

**Comment by sftim** [comment]
OK, let's _not_ squash on merge **but** can you squash this to the smallest number of changes that each represent a useful change? Doesn't have to be a squash to 1 commit.

**Comment by chalin** [comment]
Let's totally **squash it on merge** (I recommend that for all our projects). I was just hoping to keep the individual commits in this PR's branch history.

Ok, I'll look into reducing the number of commits, maybe even to one ... after I make a local clone of the branch.

**Comment by sftim** [comment]
> Let's totally **squash it on merge** (I recommend that for all our projects). I was just hoping to keep the individual commits in this PR's branch history.

Noted, but this project prefers to preserve commits from the PR.

**Comment by sftim** [comment]
Before said squash, the final commit is / was d8392ccde38b557a2e6076a7bc763aed399243b5.

**Comment by chalin** [comment]
@sftim - PR rebased and squashed to 1 commit.

File: layouts/partials/scripts.html (hunk 9 above)
Comment (chalin) [reply layouts/partials/scripts.html:25]: Yes, good candidate work in scope of:

- #41171

**Comment by sftim** [comment]
(you can fetch a commit directly from GitHub, but possibly it may eventually get garbage collected)

**Comment by nate-double-u** [comment]
I'm able to build from a fresh install both locally and using a container, but in both cases I need to do a `pushd api-ref-generator && git fetch --unshallow && popd` in order to get `make module-init` to work before things build. (I actually have do do `make module-init twice, otherwise things aren't ready for the git fetch.) 

reapplying lgtm

/lgtm

**Comment by sftim** [comment]
/hold

To make life easier, do not merge or unhold this until Kubernetes v1.32 is released. OK to unhold afterwards.
